Viewport offset aware and sticky positions

dawidurbanski opened the pull request:
This PR aims to fix contextual balloons positioning constraints if `viewportOffset` is specified.

Currently if on a page there are any absolutely positioned UI elements, some balloons can go above or under these elements. See #9892 for more information about it.

There are two subproblems to solve in this case:

1. Constrain best position finder ([`getOptimalPosition`](https://github.com/ckeditor/ckeditor5/blob/c2725effb435695c8de696f95c188eef8bf7e169/packages/ckeditor5-utils/src/dom/position.js#L80)) to only consider positions within an area limited by the newly added [`viewportOffset`](https://github.com/ckeditor/ckeditor5/blob/4620161ac67276a0d451118a3b807b66442f85d3/packages/ckeditor5-ui/src/panel/balloon/contextualballoon.js#L512) config option.
2. Add new type of position: `sticky`. To stick to the top of the constrained viewport for very tall elements and widgets (taller than the editor viewport).

It's heavily inspired by #10239 and it invalidates both #10239 and #10120.

Also, it probably will have to be slightly modified when #10352 lands in master because it independently adds the concept of `ui.viewportOffset` config property.

### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Type: TODO: I will need help with this message. There are probably a lot of breaking changes here. Closes #9892.

---

### Additional informations

Best test to check it out is:

```
yarn manual --files=ui/manual/tickets/9892/1.js
```

dawidurbanski commented:
Closed in favour of #10373

oleq commented on packages/ckeditor5-ui/src/panel/balloon/balloonpanelview.js:
```diff
@@ -719,190 +751,236 @@ BalloonPanelView.defaultPositions = {
-		name: 'arrow_ne'
-	} )
+		name: 'arrow_ne',
+		withArrow: true
```
Let's have `withArrow` defaulting to `true` (`BalloonPanelView` would do that, not `Position`) to reduce the amount of breaking changes for integrators. See my later comment in `position.js` to learn more how to deal with it.

oleq commented on packages/ckeditor5-utils/src/dom/position.js:
```diff
@@ -310,56 +238,228 @@ function getBestOfProcessedPositions( processedPositions, elementRectArea ) {
+		 * @member {String} #name
+		 */
+
+		/**
+		 * Tells whether given balloon constructed out of this position should have arrow.
+		 *
+		 * @readonly
+		 * @member {String} #withArrow
+		 */
```
Having this property here mixes the positioning logic and feature logic (the former should be feature-agnostic).

This property is relevant for `BalloonPanelView`. However, other use-cases such as `DropdownView` don't have such thing (dropdowns don't have arrows). Also, there could be other use-cases that have nothing to do with arrows and balloons: `getOptimalPosition()` should position virtually anything.

So instead of `#withArrow` let's have `#config` property (or `#presentation` which is more specific) that will pass through any configuration from the positioning function like a payload. For instance, the positioning function will be

```
	viewportStickyNorth: ( targetRect, balloonRect, viewportRect ) => {
		if ( !targetRect.getIntersection( viewportRect ) ) {
			return null;
		}

		return {
			top: viewportRect.top + BalloonPanelView.stickyVerticalOffset,
			left: targetRect.left + targetRect.width / 2 - balloonRect.width / 2,
			name: 'arrowless',
			config: {
				withArrow: false
			}			
		};
	}
```

and the config will land in `Position#config` and `BalloonPanelView` having received such an optimal position will know what to do with its config (for instance, default it to `{ withArrow: true }`).

oleq commented:
@ckeditor/qa-team Can you give us a hand with this PR? The impact of these changes is huge: we touched the mechanics of positioning balloons/popups/floatables across the entire project (also CF).

FilipTokarski commented:
What about cases when toolbar wrapping is disabled? Or when using other toolbar items when table is focused? Currently table toolbar is always on top and covers other balloons, making some options unreachable. Not sure if this is a discussion for this PR, but anyway, such cases might be annoying when editing some bigger content.
Examples below:


https://user-images.githubusercontent.com/34380544/131336802-e631a13d-10b4-41db-8b5a-f6fe5279fc34.mp4


![toolbar3](https://user-images.githubusercontent.com/34380544/131336820-b2c782b1-112b-4be8-b6aa-e379717d683d.png)

LukaszGudel commented:
### Steps:
1. Add a link somewhere inside the table.
2. Click on this link to open its balloon with the url.
3. Use arrow keys to place a caret at the end of the link.

### Result:
URL balloon will change to table toolbar, but its position will stay next to the link. This toolbar can be scrolled out of the viewport.  Changing the caret position once again with arrow keys will fix the table toolbar placement.


https://user-images.githubusercontent.com/58658048/131338042-3832d825-6131-4b20-98ec-fa1ef9691ee5.mp4

FilipTokarski commented:
### Steps:
1. Add a long table with a few empty paragraphs after it
2. Focus the table and open merge cells dropdown
3. Scroll down so that the table toolbar shows up below the table

### Expected
Merge cells dropdown is placed above the table toolbar and it's fully visible.

![toolbar6](https://user-images.githubusercontent.com/34380544/131340518-dd3d793b-1c78-4e48-843b-2a570ac3950a.png)


### Actual
Most of the times the dropdown will display properly, but if table is one of the last elements in the editor and there is enough space after it (like a few empty paragraphs), the dropdown will be hidden under the sticky toolbar, with some of its options unreachable.

https://user-images.githubusercontent.com/34380544/131340242-f3fe8f31-f888-4c4e-9d88-ea1dc3018c46.mp4

FilipTokarski commented:
All arrows disappeared from all the balloons. Was this intentional?
  
Before:

![arrow_before](https://user-images.githubusercontent.com/34380544/131345462-b3c8ae06-cf54-418e-825e-faa7f9cef34e.png)

After:

![arrow_after](https://user-images.githubusercontent.com/34380544/131345486-a4ba6b0b-2819-444d-b7f9-8378d65ac6fb.png)

oleq commented on packages/ckeditor5-utils/tests/dom/position.js:
```diff
@@ -688,8 +688,9 @@ describe( 'getOptimalPosition()', () => {
 
 function assertPosition( options, expected ) {
```
I'm missing some tests that will verify that:

*   The output of `getOptimalPosition` is an instance of `Position`
*   That `config` from the positioning function is correctly passed-through `getOptimalPosition` to the `Position` (depending whether positioning function defined it or not - see my earlier comment about the manual test).

dawidurbanski replied:
Done.

oleq reviewed (changes requested):
I found some bugs 😛 
I also refactored lots of docs in the `position.js` module because I figured it's very hard to understand the flow of data in this whole system. I created a "positioning function type" to give developers some idea of what is what and which is which. Please let me know if this makes sense to you.

oleq commented:
@Reinmar For the record, because I'm off work next week and I don't want to block the merge: I'm OK with merging this PR as soon as Dawid addresses the issues I mentioned during the last review. The solution is complete and the sooner we merge it the more time for the team to test it during this iteration. So 🤞

dawidurbanski commented on packages/ckeditor5-mention/src/mentionui.js:
```diff
@@ -463,7 +463,9 @@ export default class MentionUI extends Plugin {
 				view: this._mentionsView,
 				position: this._getBalloonPanelPositionData( markerMarker, this._mentionsView.position ),
-				withArrow: false,
+				config: {
+					withArrow: false
+				},
```
Indeed, it was some mistake on my part. Removed it completely.

dawidurbanski commented on packages/ckeditor5-ui/src/panel/balloon/balloonpanelview.js:
```diff
@@ -244,9 +245,11 @@ export default class BalloonPanelView extends View {
 
-		Object.assign( this, { top, left, position } );
+		const { name: position, config = {} } = optimalPosition;
+		const { withArrow } = config;
```
Fixed.

dawidurbanski commented on packages/ckeditor5-ui/src/panel/balloon/balloonpanelview.js:
```diff
@@ -420,11 +423,26 @@ BalloonPanelView.arrowHorizontalOffset = 25;
 BalloonPanelView.arrowVerticalOffset = 10;
 
+/**
+ * A vertical offset of the balloon panel from the edge of the viewport if sticky.
+ * It helps in accessing toolbar buttons underneath the balloon panel.
+ *
+ *       Toolbar                        /-- vertical offset
+ *       +-----------------------------V--------------------+
+ *       |                  +-------------+                 |
+ *       +------------------|   Balloon   |-----------------+
+ *                          +-------------+
```
Updated.

```
 *		  +---------------------------------------------------+
 *		  |                      Target                       |
 *		  |                                                   |
 *		  |                            /-- vertical offset    |
 *		+-----------------------------V-------------------------+
 *		| Toolbar            +-------------+                    |
 *		+--------------------|   Balloon   |--------------------+
 *		| |                  +-------------+                  | |
 *		| |                                                   | |
 *		| |                                                   | |
 *		| |                                                   | |
 *		| +---------------------------------------------------+ |
 *		|                        Viewport                       |
 *		+-------------------------------------------------------+
```

dawidurbanski commented on packages/ckeditor5-ui/tests/panel/balloon/balloonpanelview.js:
```diff
@@ -969,6 +980,59 @@ describe( 'BalloonPanelView', () => {
+			} );
+		} );
+
+		it( 'should stick position to the top when target element is above viewport', () => {
```
Actually It's covered in this test: https://github.com/ckeditor/ckeditor5/blob/ck/9892-viewport-top-offset-balloons-v3/packages/ckeditor5-ui/tests/panel/balloon/balloonpanelview.js#L986

But I agree it's not so obvious so I'll ad another one.

dawidurbanski replied:
Added.

dawidurbanski commented on packages/ckeditor5-ui/tests/panel/balloon/balloonpanelview.js:
```diff
@@ -969,6 +980,59 @@ describe( 'BalloonPanelView', () => {
+			} );
+		} );
+
+		it( 'should stick position to the top when target element is above viewport', () => {
```
Updated.

dawidurbanski commented:
@FilipTokarski @LukaszGudel Thanks for looking at this. Just to make sure I addressed all the issues you raised:

1. In this [issue](https://github.com/ckeditor/ckeditor5/pull/10373#issuecomment-908294152) you do touch a topic of balloons collisions. It's something we can improve for sure, but we do not care about it at this point. Making collision detection and global balloons auto-positioning mechanism is something we maybe will tackle in the future.

2. In this [issue](https://github.com/ckeditor/ckeditor5/pull/10373#issuecomment-908298028) it's caused by a mechanism that prevents balloons from jumping around if the context has been changed (for example if you are on a link, but now you jumped out of the link and you see text attributes balloon. It should stay more or less in the same place). Because of this the balloon stays inside the cell, and cell related balloons (for example table cell properties) do not get sticky when scrolling (because it makes no sense for the UI). So we decided it's good enough as is.

3. About this [issue](https://github.com/ckeditor/ckeditor5/pull/10373#issuecomment-908310144). After quick discussion with @oleq we decided it's an edge case we don't want to bother with.

4. This [issue](https://github.com/ckeditor/ckeditor5/pull/10373#issuecomment-908338011) with missing arrow has been fixed.


@ckeditor/qa-team Could you quickly re-test this thing with everything I posted above in mind?

Mgsy commented:
LGTM :+1:

Reinmar commented:
Thank you all :) It's been a long ride, but an interesting one. I scanned most of the code changes and they look really nice. I was surprised, but I didn't actually spot breaking changes except one (in the widget package). The changes in utils seem to be backwards compat because the new Position class is compatible with the old typedef and the `getOptimalBleble()` will work without the new option. Please check my merge commit message and add missing info (with an empty commit to `master`) if I missed something.

Once again, thank you and congrats 👏
